### ISSUE TITLE
[css-nesting-1] Clearer definition of “nest-containing”

### ISSUE THREAD

**Issue by sesse** [w3c/csswg-drafts#7972]
The spec currently reads: “If a selector in the [<relative-selector-list>](https://csswg.sesse.net/selectors-4/#typedef-relative-selector-list) does not start with a [combinator](https://csswg.sesse.net/selectors-4/#selector-combinator) but does contain the [nesting selector](https://csswg.sesse.net/css-nesting-1/#nesting-selector), it is interpreted as a non-[relative selector](https://csswg.sesse.net/selectors-4/#relative-selector).”

But it doesn't really say what “contain” means. An example we discussed earlier was:

`.foo { :is(.bar, !#&/) { … }}`

Does this “contain” the nesting selector? It has a & in there, but it's thrown away due to parse error (and then forgiving parsing should kick in and retain only `.bar`.). (This is a contrived example, of course, but you could imagine something like `:is(.bar, & >> .foo)` for a hypothetical `>>` combinator. A newer browser would interpret it as selector-containing, an older that did not understand `>>` would not.)

I assume this _doesn't_ contain the nesting selector:

`.foo { :is(/*fluid&flexible*/.bar) { … }}`

but it's difficult to distinguish from the previous case.

Finally, there's also

`.foo { :is(& .bar, .bar) { …}}`

where one side of the :is() contains a nesting selector, which I'd interpret as the entire selector doing so, but it's not 100% obvious.

The simplest thing for us would certainly be to let “contain” mean a properly parsed & somewhere, so the answers here would be no, no, yes, respectively. But if you changed the third case to “no” (requring all arms of the :is() to have & for the :is to be selector-containing), you would probably avoid differences in cases like the first one between old and new browsers.

**Comment by bradkemper** [comment]
I think each complex selector in a selector list should be considered separately. So it one is garbage and is thrown out of a forgiving function, then it doesn't affect the other.

**Comment by sesse** [comment]
So is `:is(.bar, &)` to be interpreted as a relative selector or not?

**Comment by bradkemper** [comment]
This is just my opinion, but I would say both. For the purposes of determining whether there is an implicit `&` (but still basing specificity on the most specific part), it would be as though that was divided into two rules, like this:

```
:is(.bar) {...}
:is(&) {...}
```
So then, this:

```
& :is(.bar) {...}
:is(&) {...}
```

**Comment by bradkemper** [comment]
So then, to bring back to a single selector list 

```

:is(&.bar, &) {...}

```

Not sure if that works consistently or not; I probably haven't considered it enough.

**Comment by sesse** [comment]
So again, what about this? `:is(.bar, &:thispseudodoesnotexist)`

**Comment by bradkemper** [comment]
```

:is(& .bar, &:thispseudodoesnotexist)

```

And since `:is()` is forgiving, then becomes this:

```

:is(& .bar)

```

which is equivalent to:

```

& .bar

```

**Comment by sesse** [comment]
I don't think this is going to work; just adding `&` everywhere is going to be increasingly strange and not very useful. (Also, the standard right now is defined in terms of relative selectors, which don't work like this at all.) For instance, `:is(.a) > .is(.b)` should then become `:is(& .a) > :is(& .b)`? Which means that `:is(.a) > .is(.b)` becomes something entirely different from `.a > .b` (which according to the standard should become `& .a > .b`)? Or that `:is(:is(.a))` should become `:is(& :is(& .a))`?

**Comment by tabatkins** [comment]
Proposal Part 1: Selector Lists
--------------------------------

Split on top-level commas, and decide whether each selector is relative or not independently. (`.foo { .bar, .baz & {...}}` desugars to `.foo .bar, .baz .foo {...}`). (Doing anything else seems like it would give strange results, and result in worse editting hazards for authors.)

Proposal Part 2: Enclosed &
--------------------------------

* Option 1: Relativeness is decided based on only top-level &. An & inside any other construct (such as `:is()` or `:nth-child()`) is ignored for this purpose.
* Option 2: Relativeness is decided based on & both at the top level and within functional syntaxes (such as `:has()` or `:is()`).
* Option 3: Relativeness is decided based on & both at the top level and within only *non-forgiving* functional syntaxes (visible in `:has()` or `:nth-child()` but ignored in `:is()` or `:where()`).

Option 2 seems to be the most preferable to me, but it's unclear how `:is(:unknown(&), .bar)` should (or even *can be*) handled - once we're no longer looking at a valid selector, all bets are off on what that `&` might mean. It might be reasonable to assume we won't use `&` in a selector context in a way that doesn't represent the nesting selector, but `:is()` drops its invalid arguments when serialized, so reserializing could change the relative/non-relativeness of the selector which is bad.

Between Option 1 and Option 3, Option 1 has the advantage of being simpler to understand and explain. Both of these can potentially lead to selectors that simply cannot be written in a nested fashion (for example, `.foo { :not(&) {...}}`), but the solution is to just... not nest those. And the simpler and easier we make such situations to notice and understand, the better.

So, I propose we go with Option 1, and only recognize & at the top-level of a selector to determine whether the nested selector is relative or not.

**Comment by sesse** [comment]
Option 2 seems fine to me; I don't honestly think people will be writing  `:is(:unknown(&), .bar)` too much in the first place (surely, `:is(:unknown(&), & .bar)` would be a more common case?). We have to give up something, and between handling forward-compatibility and distinguishing between `&` inside and and outside of parens, the latter seems more broadly useful to me. (“Just don't nest” is limiting the standard unneccessarily, I'd feel? As long as we've said nesting is something fundamentally different from `@scope` and more like a grouping handling similar to a macro, we should probably accept these cases as equally valid.)

**Comment by tabatkins** [comment]
Yeah, after discussing with @nex3 (the situation doesn't apply to Sass, since it doesn't have forgiving parsing anywhere), I do lean towards Option 2 as well. To handle the rare weird case where it will matter, I'll probably have to specify some sort of "guaranteed never-matching selector" that can be paired with &, and specify that an `:is()` has to insert that if it would drop an invalid selector that contains an `&` token and none of the remaining valid selectors have it, but that's a corner case we don't need to worry about imminently.

(That is, `:is(:unknown(&), .bar)` would serialize to `:is(&:matches-nothing, .bar)` instead, but `:is(:unknown(&), & .bar)` would just serialize to `:is(& .bar)` as normal.)

**Comment by romainmenke** [comment]
> `:is(:unknown(&), .bar)`

```css
.foo {
  :is(:unknown(&), .bar) {}
}
```

I might be overlooking this but what happens to the `.bar` branch?
The selector as a whole is not relative because it contains `&`.

I assume it is a valid selector that never matches anything?
I would be surprised if it matches any `.bar` in the document, regardless of elements having a `.foo` ancestor.

**Comment by tabatkins** [comment]
It does indeed match any `.bar` element (that is, it's equivalent to `.foo .any, .bar {...}`). I agree it feels a little odd (since it's the only way to write such a selector in a nested context), but the behavior isn't totally off-base. After all, `.foo { :not(&) {...}}` selects all the elements on the page that aren't `.foo`, also with absolutely no regard to whether they're in a `.foo` ancestor or not. Nesting isn't `@scope` ^_^

**Comment by romainmenke** [comment]
Actually happy it works this way :)
It is a bit odd indeed, but easy to understand if you don't overthink it.

**Comment by sesse** [comment]
We're adjusting the Blink implementation now, but be aware that this breaks round-tripping. `:is(!&, .c)` is nest-containing, but if you serialize it, it becomes `:is(.c)`, which isn't.

**Comment by cdoublev** [comment]
> `& appearing *anywhere* in the selector`

Sorry, but I am confused.

According to https://github.com/w3c/csswg-drafts/issues/8310#issuecomment-1383215582, `&` must replace a `<simple-selector>`.

According to https://github.com/w3c/csswg-drafts/issues/7503#issuecomment-1377827748, `&` can replace a sub-production of `<simple-selector>`.

According to the above comment, it can match anything? `!&` is not a valid `<relative-selector>`. 

The selector of the nested style rule is nest-containing if it contains `&` even as part of an invalid `<relative-selector>`? And in a comment? Or escaped (thus matching `<ident>`, thus matching `<type-selector>`)?

**Comment by tabatkins** [comment]
@sesse Note the second resolution in the minutes - if a forgiving-selector-list entry contains an &, we *don't* drop it when serializing. I'll be updating the spec with this text in a little bit.

> Sorry, but I am confused.

Yup, Oriol got it. We look for delim tokens, and this doesn't change the existing definition of the nesting selector; it's something unrelated happening at an earlier stage of parsing.

**Comment by tabatkins** [comment]
All right, edited in with the resolutions from above:

* a selector "contains the nesting selector" if, during parsing, it encoutners a delim-token with the value "&" *anywhere*
* a monkey-patch of the forgiving-selector-list behavior, where it must preserve list items that contain the nesting selector

**Comment by cdoublev** [comment]
> * a monkey-patch of the forgiving-selector-list behavior, where it must preserve list items that contain the nesting selector

Can you please clarify if `<forgiving-selector-list>` should preserve all list items, even those that do not contain a nesting selector and are invalid?

> If a `<forgiving-selector-list>` has an item that contains the nesting selector but is invalid, **it** is preserved exactly as-is rather than being discarded.

Emphasize added on "it" because it is not clear to me if it refers to "item" or `<forgiving-selector-list>`.

Maybe an example would be clearer than refining the prose.
